# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0304 Карабанов Роман, вариант 35.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.


## Основные теоретические положения

Аддитивной цепочкой для $ n \in \mathbb{N} $ называется последовательность натуральных чисел
$$ 1 = a_0, a_1, ..., a_r = n, $$
где каждый элемент последователньости равен сумме каких-то двух предыдущих:
$$ a_i = a_j + a_k, \quad k \le j \le i, \quad i = 1, 2, ..., r $$

$l(n) = r$ - наименьшая длина аддитивной цепочки для $ n \in \mathbb{N} $.

Для метода множителей: $ \quad l(mn) \le l(m) + l(n)$

Для бинарного метода: $l(n) = \lambda(n) + \nu(n) - 1, \quad$ где $\lambda (n) = \lfloor lb(n) \rfloor, \nu(n)$ - вес Хэмминга числа $n$ (количество единиц в двоичной записи).

Для m-арного метода: $ \quad l(n) \le m - 2 + (k + 1)t, \quad$ где $m = 2^k, n = \sum_{j = 0}^t d_j m^{t-j} $

Для SX-метода: $\quad l(n) \le \lambda (n) + \nu (n) - 1$


Пара $ (j, k), 0 \le k \le j < i $ называется **шагом** $i$.Если существует более одной пары $(j, k)$, полагаем $j$ наибольшим.

## Виды шагов:
* удвоение: $ j = k = i - 1 $
* звёздный шаг: $ j = i - 1 $ (линейный шаг)
* малый шаг: $ \lambda (a_i) = \lambda (a_i - 1) $

## Свойства видов шагов:
* шаг 1 - всегда удвоение
* удвоение - звёздный шаг, но никогда не малый
* если $i$-ый шаг не малый, то $(i+1)$-ый шаг либо малый, либо звёздный, либо оба
* за удвоением всегда следует звёздный шаг
* если $(i+1)$-ый шаг не звёздный и не малый, то $i$-ый шаг должен быть малым

## Теорема:

Если аддитивная цепочка содержит $d$ и $f = r - d$ неудвоений, то $n \le 2^{d-1} F_{f+3}$, где $F_j$ - число Фиббоначи

## Следствие:

Если аддитивная цепочка содержит $f$ удвоений и $S$ малых шагов, то

${S \le f \le} {S \over {1 - lb(\varphi)}}$, где $\varphi = {{\sqrt{5} + 1} \over 2} $ - золотое сечение

### Алгоритм Брауэра:

Для $n \in \mathbb{N}$ при заданном $k \in \mathbb{N}$ можно построить цепочку Брауэра с помощью рекуррентной формулы:

$$ B_k (n) =
\begin{cases}
1, 2, 3, ..., 2^k - 1, \quad n < 2^k \\
B_k (q), 2q, 4q, ..., 2^k q, n, \quad n \ge 2^k
\end{cases} \\
q = \lfloor {n \over 2^k} \rfloor
$$

Длина цепочки
$$l_B (n) = j(k + 1) + 2^k - 2,$$
при условии, что $jk \le \lambda (n) \le (j+1)k$

Длина будет минимальной для больших $n$, если положить $k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n)$

## Ход алгоритма:

* Задаётся некий параметр $k$ для $n$.
Вычисляются вспомогательные числа:
$$
d = 2^k, \hspace{0.2cm} q_1 = [ {n \over d} ], \hspace{0.2cm} r_1 = n \hspace{0.2cm} mod \hspace{0.2cm} d => n = q_1 d + r_1 \quad (0 \le r_1 < d) \\
q_2 = [ {q_1 \over d} ], \hspace{0.2cm} r_2 = q_1 \hspace{0.2cm} mod \hspace{0.2cm} d => q_1 = q_2 d + r_2 \quad (0 \le r_2 < d)
$$
* Данная процедура продолжается, пока не появится $q_s < d.$ Следовательно, $q_{s-1} = q_s d + r_s$
* Таким образом, n имеет вид
$$ n = 2^k q_1 + r_1 = 2^k (2^k q_2 + r_2) + r_1 = ...\\
... = 2^k (2^k (... (2^k q_s + r_s ) ... ) + r_2 ) + r_1 . $$

$$B_n (n): 1, 2, 3, ..., 2^k - 1, \\
2q_s, 4q_s, 8q_s, ..., 2^k q_s, 2^k q_s + r_s, \\
2q_{s-1}, 4q_{s-1}, 8q_{s-1}, ..., 2^k q_{s-1}, 2^k q_{s-1} + r_{s-1}, \\
..., \\
..., 2^k q_1, 2^k q_1 + r_1 = n.$$


## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных
аддитивных цепочек с использованием системы компьютерной математики
SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.
## Порядок выполнения работы
1. Вручную (т.е. не реализовывая алгоритм на Sage) построить последовательность
вычислений бинарным методом и методом множителей
для $x^n$ для 2-3 значений $n$ (значение $n > 30$ выбираются
студентом самостоятельно). Сравнить количество операций для
каждого метода, сделать выводы.
2. Реализовать алгоритм Брауэра (для нечётных вариантов) и алгоритм
Яо (для чётных вариантов) для вычисления приближённых
аддитивных цепочек для различных чисел при варьировании
параметра $k$, сопоставить длины полученных аддитивных цепочек
с минимальной аддитивной цепочкой для заданного числа. Сделать
выводы.
3. Реализовать алгоритм дробления вектора индексов для нахождения
минимальной звёздной цепочки для заданного числа. Протестировать
алгоритм минимум для 5 значений $n > 1000$. Указать,
сколько времени потребовалось на поиск цепочки и какая цепочка
получилась. Сравнить с предыдущими методами, сделать выводы.
4. Проверить гипотезу Шольца–Брауэра для всех натуральных
$1 \le n \le 12$ на алгоритме дробления вектора индексов. Результаты
оформить в виде таблицы. Сделать выводы.
5. Найти или предложить собственные модификации алгоритмов и
привести описание модификаций. Реализовать модифицированные
алгоритмы и сравнить их мощность.


## Выполнение работы
### Задание 1
####  Выберем числа n = 33, 125
#### 1.1 Бинарный метод.
Для числа n = 33 аддитивная цепочка будет выглядеть:
$$ a_0 = 1, 2, 4, 8, 16, 32, 33 = n $$
То есть для $x^n$, n = 33 число операций бинарным методом будет = 6

Для числа n = 125 аддитивная цепочка будет выглядеть:
$$ a_0 = 1, 2, 3, 6, 7, 14, 15, 30, 31, 62, 124, 125 = n$$
То есть для $x^n$, n = 64 число операций бинарным методом будет = 11

#### 1.1 Метод множителей.
Для числа n = 33 

$x^{33} = (x^{11})^3, (y^3 = y^2 * y, y^2 = y * y)$

$x^{11} = x^{10} * x$

$x^{10} = (x^5)^2$

$x^5  = x^4 * x$

$x^4 = x^2 * x^2$

$x^2 = x * x$

Отсюда требуемые операции : $ x, x^2, x^4, x^5, x^{10}, x^{11}, x^{22}, x^{33} $ = 7

Для числа  = 125

$x^{125} = (x^{25})^5, (y^5 = y^4 * y, y^4 = y^2 * y^2, y^2 = y * y)$

$x^{25} = (x^5)^5, (y^5 = y^4 * y, y^4 = y^2 * y^2, y^2 = y * y)$

$x^5 = x^4 * x $

$x^4 = x^2 * x^2 $

$x^2 = x * x$

Отсюда требуемые операции : $ x, x^2, x^4, x^5, x^{10}, x^{20}, x^{25}, x^{50}, x^{100}, x^{125} $ = 9

### Вывод на основе результов используемых значений
Таким образом видно, что для числа 33 метод множителей даёт результат по операция больший, чем бинарный метод, т.к он работает "медленне" по отношению к бинарному методу при степенях $ n =  2^k + 1 $ в то время, как для числа 125 метод множителей даёт меньшее количество операций, по сравнению с бинарным методом. Можно сделать выводы о том, что количество операций, необходимых для метода множителей равно сумме длин аддитивных цепочек сомножителей, поэтому ни один из методов не может рассматриваться как более эффективный.

### Задание 2
### Реализация алгоритма Брауэра
Реализованный алгоритм Брауэра представлен, с вариацией значений k представлен ниже:

In [39]:
def alg(n, k):
      d = 2 ** k
      ans = []
      if (n < d):
        i = 0
        while i < n:
            ans.append(i + 1)
            i += 1
      else:
          q = n
          ans_r = []
          while q >= d:
              r = q % d
              ans_r.append(r)
              q = q // d
          ans_r.append(q)
          ans_r.reverse()
          z = 1
          while z < d:
              ans.append(z)
              z += 1          
          q = ans_r[0]
          i = 0
          while i < (len(ans_r) - 1):
              j = 0
              while j < k:
                  q *= 2
                  ans.append(q)
                  j += 1
              if ans_r[i+1] > 0:
                  q += ans_r[i+1]
                  ans.append(q)
              i += 1
      return ans, (len(ans))

n = int(input("n: "))
k = int(input("k: "))
ad = []
lng = 0
ad, lng = alg(n, k)
print("Аддитивная цепочка: ", ad)
print("Длина аддитивной цепочки: ", lng)


n: 127
k: 10
Аддитивная цепочка:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127]
Длина аддитивной цепочки:  127


Для $ n = 90 $ результаты:

| n  | k | минимальная длина | длина | аддитивная цепочка                                 |
|----|---|-------------------|-------|----------------------------------------------------|
| 90 | 2 | 9                 | 12    | 1,2,3,4,5,10,20,22,44,88,90                        |
| 90 | 3 | 9                 | 15    | 1,2,3,4,5,6,7,2,4,8,11,22,44,88,90                 |
| 90 | 4 | 9                 | 20    | 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,10,20,40,80,90 |

Для $ n = 512$ результаты:

| n  | k | минимальная длина | длина | аддитивная цепочка                                 |
|----|---|-------------------|-------|----------------------------------------------------|
| 512 | 2 | 11                 | 11    | 1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512                       |
| 512 | 3 | 11                 | 16    | 1, 2, 3, 4, 5, 6, 7, 2, 4, 8, 16, 32, 64, 128, 256, 512                 |
| 512 | 4 | 11                 | 20    | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 4, 8, 16, 32, 64, 128, 256, 512 |

Для $ n = 127 $ результаты:

| n  | k | минимальная длина | длина | аддитивная цепочка                                 |
|----|---|-------------------|-------|----------------------------------------------------|
| 127 | 2 | 12                 | 12    | 1, 2, 3, 2, 4, 7, 14, 28, 31, 62, 124, 127                        |
| 127 | 3 | 12                 | 15    | 1, 2, 3, 4, 5, 6, 7, 2, 4, 8, 15, 30, 60, 120, 127              |
| 127 | 10 | 12                 | 127    | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ... , 127 |

Таким образом, можно сделать вывод, что алгоритм Брауэра является приближенным и при определённых значениях $ k $ выдаёт цепочки близкие или совпадающие с минимальными, но при больших k длина аддитивной становится во много раз больше минимальной (последний пример)


### Задание 3
### Алгоритм дробления вектора индексов




### Задание 4
### Проверить гипотезу Шольца–Брауэра для всех натуральных  $1≤𝑛≤12$  на алгоритме дробления вектора индексов.


## Вывод